### PR TITLE
Notification system polish: system toasts, sounds, bubbling, dock badge

### LINKED ISSUE #21: Notification system polish: system notifications, ID robustness, delivery audit
## Context

The in-app notification system works (OSC 9/99/777 parsing, notification panel, sidebar badges, pane flash rings), but it's missing system-level notifications entirely and has some areas worth hardening around ID generation and delivery routing.

## 1. System notifications (macOS / Windows / Linux)

amux currently only has in-app notifications. cmux uses `UNUserNotificationCenter` to deliver native macOS notifications when the app is unfocused, with the notification ID as the system notification identifier so clicking it can jump to the right tab.

We should add native notification delivery when the amux window is not focused:
- **macOS**: `UNUserNotificationCenter` (or `notify-rust` crate which wraps it)
- **Windows**: Toast notifications via `notify-rust` or `winrt`
- **Linux**: D-Bus `org.freedesktop.Notifications` via `notify-rust`

`notify-rust` covers all three platforms and would be the simplest path. Clicking a system notification should focus the amux window and jump to the relevant workspace/pane.

## 2. ID generation: sequential counters vs UUIDs

amux uses sequential `u64` counters (`next_pane_id`, `next_workspace_id`, `next_surface_id`) that persist across sessions. cmux uses `UUID()` for all entity and notification IDs.

**Current state**: The sequential approach works correctly in normal operation — counters are persisted in `session.json` and continue monotonically across restarts. No ID collisions have been observed.

**Potential concern**: If the session file is corrupted, manually edited, or fails to save (crash during write), counters could reset or duplicate. The atomic write (write to `.tmp` then rename) in `amux-session` mitigates the crash case, but doesn't cover:
- Session file deleted manually → counters restart at 1
- Two amux instances launched simultaneously (unlikely but possible)

**Assessment**: The sequential counter approach is fine for now. The persisted counters plus atomic writes make collisions extremely unlikely. UUIDs would eliminate the theoretical risk but add complexity for debugging (sequential IDs are much easier to reason about in logs). **No change recommended unless we hit a real bug.**

## 3. Notification delivery audit

The delivery path (`drain_notifications()` in main.rs) routes notifications based on focus state at drain time:
- Focused pane → `push_read()` (flash only, no unread badge)
- Unfocused pane → `push()` (flash + unread badge + sidebar count)

**Edge case to investigate**: If focus changes between when an OSC notification event fires and when `drain_notifications()` runs in the next frame, a notification could be misclassified. This is low probability (single-threaded event loop, drain runs every frame) but worth adding a regression test for.

**Workspace lookup**: `workspace_for_pane()` iterates all workspaces to find which one owns a pane. This is O(workspaces × panes) per notification. Fine for typical use but could be optimized with a reverse lookup map if it ever becomes a bottleneck.

## Suggested approach

1. Add `notify-rust` dependency and gate behind a `system-notifications` feature flag
2. In `drain_notifications()`, when the window is unfocused and a new unread notification is pushed, also fire a system notification
3. Handle system notification click → focus window + jump to workspace/pane
4. Add integration test for focus-change-during-drain edge case
5. Leave ID generation as sequential counters (document the rationale)

## Reference

- cmux system notifications: `Sources/TerminalNotificationStore.swift:1056-1092`
- amux notification store: `crates/amux-notify/src/lib.rs`
- amux notification delivery: `crates/amux-app/src/main.rs` (`drain_notifications()`)
- amux ID generation: `crates/amux-app/src/main.rs` (lines 936-938)

### PR DESCRIPTION
**Pull request by daveowenatl**
## Summary

Implements cmux-parity notification features (Issue #21):

- **Notification config**: Expand `AppConfig` with `[notifications]` section covering system_notifications, sound, auto_reorder_workspaces, dock_badge, custom_command
- **Window focus tracking**: Track `app_focused` via egui's `InputState::focused` for notification suppression
- **Workspace bubbling**: Auto-reorder workspaces to top of sidebar when they receive notifications (matching cmux's `moveTabToTopForNotification`)
- **System notifications**: OS-native toast notifications via `notify-rust` (UNUserNotificationCenter on macOS, Windows toast notifications) with click-to-navigate action channel
- **Three-tier delivery**: Matches cmux's model — unfocused (OS toast + command), focused-different-pane (in-app sound + command), focused-same-pane (nothing)
- **Notification sounds**: Cross-platform audio via `rodio` (CoreAudio/WASAPI/ALSA) with "system" beep, "none", or custom audio file modes
- **Custom command**: Shell command on notification with `AMUX_NOTIFICATION_{TITLE,BODY,SOURCE}` env vars
- **Dock badge**: macOS dock tile badge label with total unread count (capped at "99+") via objc2-app-kit

## New dependencies
- `notify-rust` v4 — cross-platform system notifications
- `rodio` v0.20 — cross-platform audio playback (wav/ogg/mp3)
- `objc2-foundation` v0.3 + `objc2-app-kit` v0.3 — macOS dock badge (already transitively available)

## Config example
```toml
[notifications]
system_notifications = true
auto_reorder_workspaces = true
dock_badge = true
custom_command = "say 'notification received'"

[notifications.sound]
sound = "system"  # "system", "none", or "/path/to/sound.wav"
play_when_focused = true
```

## Test plan
- [ ] Launch amux, open two workspaces, trigger notification in workspace 2 → verify it bubbles to top
- [ ] Minimize amux, trigger notification → verify OS toast appears
- [ ] Focus amux on workspace 1, trigger notification in workspace 2 → verify in-app sound
- [ ] Verify dock badge shows unread count, clears when read
- [ ] Set `sound = "none"` → verify silent
- [ ] Set `auto_reorder_workspaces = false` → verify no bubbling
- [ ] `cargo build --no-default-features` (soft renderer fallback)

Closes #21

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * System notifications with clickable actions to navigate to workspaces.
  * Configurable notification sound playback (system, none, or custom file).
  * Dock/taskbar badge shows unread counts (with cap/clear behavior).
  * Smart delivery: OS toasts when app unfocused, in-app sounds when focused.
  * Optional custom command execution on notifications.
  * IPC-triggered notifications and automatic workspace "bubbling" to top.

* **Tests**
  * Added unit test verifying total unread count across panes.

* **Chores**
  * Added platform audio/notification libraries and macOS integrations; CI installs Linux audio packages.

### CODE REVIEW

File: crates/amux-app/src/system_notify.rs [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+            if let Err(e) = notification.show() {
+                tracing::warn!("Failed to show system notification: {}", e);
+            }
+        });
+    }
+
+    /// Drain any click-to-navigate actions from notification callbacks.
+    pub fn drain_actions(&self) -> Vec<NotificationAction> {
+        self.action_rx.try_iter().collect()
+    }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-app/src/system_notify.rs:136]: Fixed. Documented that drain_actions() is scaffolding — notify-rust doesn't expose cross-platform click callbacks. The action channel is plumbed for when we add platform-specific click handling (or fall back to spawning `amux focus` as a subprocess). Added clear doc comment on both the struct and drain_actions().

File: crates/amux-app/src/system_notify.rs [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    #[cfg(target_os = "windows")]
+    {
+        // Windows: flash the taskbar button when there are unread notifications.
+        // A full count overlay via ITaskbarList3::SetOverlayIcon is more complex
+        // and can be added later.
+        if count > 0 {
+            // FlashWindowEx would go here, but requires the HWND.
+            // For now, this is a placeholder — eframe doesn't expose HWND directly.
+            let _ = count;
+        }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-app/src/system_notify.rs]: Fixed. Updated docstring to accurately reflect Windows is currently a no-op placeholder (requires HWND access not yet available from eframe).

File: crates/amux-app/src/system_notify.rs [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+                notification.body(&body);
+            }
+
+            // On macOS, notify-rust uses UNUserNotificationCenter which
+            // handles permission prompts automatically on first use.
+            if let Err(e) = notification.show() {
+                tracing::warn!("Failed to show system notification: {}", e);
+            }
+        });
+    }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-app/src/system_notify.rs]: Fixed. Replaced per-notification thread spawning with a single long-lived worker thread (`amux-notify-worker`) that processes notifications and custom commands sequentially via an mpsc channel.

File: crates/amux-app/src/system_notify.rs [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+            .stdout(std::process::Stdio::null())
+            .stderr(std::process::Stdio::null())
+            .spawn()
+        {
+            Ok(_) => {}
+            Err(e) => {
+                tracing::warn!("Failed to run notification command: {}", e);
+            }
+        }
+    });
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-app/src/system_notify.rs]: Fixed. Custom commands now route through the same background worker thread as system notifications, eliminating unbounded thread creation.

File: crates/amux-app/src/system_notify.rs [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+                                .play_raw(rodio::Source::convert_samples(source))
+                            {
+                                tracing::warn!("Failed to play custom sound: {}", e);
+                            }
+                        }
+                        Err(e) => {
+                            tracing::warn!("Failed to decode custom sound: {}", e);
+                        }
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-app/src/system_notify.rs:243]: Fixed. Changed custom_sound from `Option<Vec<u8>>` to `Option<Arc<[u8]>>`. Play now clones the Arc (cheap refcount bump) and uses a custom `ArcSliceReader` implementing Read+Seek for rodio.

File: crates/amux-app/src/main.rs [hunk 6]
```diff
@@ -1253,6 +1320,27 @@ impl eframe::App for AmuxApp {
...
+                    ws.focused_pane = action.pane_id;
+                }
+                self.notifications.mark_pane_read(action.pane_id);
+            }
+        }
+
+        // Update dock/taskbar badge with total unread count
+        if self.app_config.notifications.dock_badge {
+            system_notify::set_badge_count(self.notifications.total_unread());
+        }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-app/src/main.rs]: Fixed. Added `last_badge_count: usize` to AmuxApp and only call `set_badge_count()` when the count actually changes.

File: crates/amux-app/src/main.rs [hunk 7]
```diff
@@ -1253,6 +1320,27 @@ impl eframe::App for AmuxApp {
...
+                .position(|ws| ws.id == action.workspace_id)
+            {
+                self.active_workspace_idx = idx;
+                let ws = &mut self.workspaces[idx];
+                if ws.tree.iter_panes().contains(&action.pane_id) {
+                    ws.focused_pane = action.pane_id;
+                }
+                self.notifications.mark_pane_read(action.pane_id);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-app/src/main.rs]: Acknowledged. The drain_actions() wiring is intentional scaffolding for future click-to-navigate support. Added doc comments clarifying this. Once we implement platform-specific click handling, we'll also add window focus/foreground behavior as you suggest.

File: crates/amux-app/src/system_notify.rs [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+
+/// Cross-platform system notification sender using `notify-rust`.
+///
+/// Sends OS-native toast notifications (UNUserNotificationCenter on macOS,
+/// Windows toast notifications on Windows). Notification clicks are routed
+/// back via an mpsc channel as `NotificationAction`s.
+pub struct SystemNotifier {
+    action_rx: mpsc::Receiver<NotificationAction>,
+    _action_tx: mpsc::Sender<NotificationAction>,
+}
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-app/src/system_notify.rs:39]: Acknowledged. The action channel is intentional scaffolding for future click-to-navigate support. notify-rust doesn't currently expose cross-platform click callbacks. Added clear documentation in the previous commit. Will route through set_focus() when we implement the callback handling.

File: crates/amux-app/src/system_notify.rs [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+            .env("AMUX_NOTIFICATION_SOURCE", &source)
+            .stdout(std::process::Stdio::null())
+            .stderr(std::process::Stdio::null())
+            .spawn()
+        {
+            Ok(_) => {}
+            Err(e) => {
+                tracing::warn!("Failed to run notification command: {}", e);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-app/src/system_notify.rs]: Fixed. Changed .spawn() to .status() and added stdin(Stdio::null()). Since this runs on the worker thread (not UI thread), waiting for the child is fine and properly reaps the process.

**Comment by daveowenatl** [comment]
## CodeRabbit review-body findings addressed

### Outside diff range — Critical: Background notifications suppressed for focused pane (lines 2768-2802)
**Fixed.** Restructured `drain_notifications()` to check `self.app_focused` first. When app is unfocused, all notifications are treated as background (OS toast + command + unread), regardless of which pane they target. The `pane_id == focused` check now only applies when the app is focused (tier 3).

### Outside diff range — Major: `notify.send` bypasses three-tier delivery (lines 4563-4582)
**Fixed.** Applied the same three-tier delivery logic to the IPC `notify.send` handler:
- Tier 1 (unfocused): system toast + command + unread
- Tier 2 (focused, different pane): sound + command + unread
- Tier 3 (focused, same pane): push_read (flash only)

### Nitpick: Cargo feature for notifications (Cargo.toml lines 33-38)
**Deferred.** The notification dependencies are relatively lightweight and always desired in normal builds. A compile-time feature gate adds complexity for a marginal benefit since the runtime config already gates all behavior. Will revisit if build times become an issue.

File: crates/amux-app/src/system_notify.rs [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+#[derive(Clone)]
+enum SoundMode {
+    System,
+    None,
+    Custom,
+}
+
+/// A short 440Hz sine wave beep (~150ms) generated at runtime.
+fn system_beep_source() -> rodio::source::SineWave {
+    rodio::source::SineWave::new(440.0)
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-app/src/system_notify.rs:162]: Not a bug — rodio 0.20's `SineWave::new` takes `f32`, not an integer. Verified: `pub fn new(freq: f32) -> SineWave` in rodio-0.20.1/src/source/sine.rs:21. The `440.0` literal is correct.

File: crates/amux-app/src/system_notify.rs [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+pub struct SystemNotifier {
+    action_rx: mpsc::Receiver<NotificationAction>,
+    _action_tx: mpsc::Sender<NotificationAction>,
+    worker_tx: mpsc::Sender<WorkerMsg>,
+}
+
+impl SystemNotifier {
+    pub fn new() -> Self {
+        let (action_tx, action_rx) = mpsc::channel();
+        let (worker_tx, worker_rx) = mpsc::channel::<WorkerMsg>();
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-app/src/system_notify.rs]: Fixed. Switched to `mpsc::sync_channel::<WorkerMsg>(100)` with bounded capacity. This provides backpressure during notification bursts — the sender will block if the worker falls behind.

File: crates/amux-notify/src/lib.rs [hunk 12]
```diff
@@ -316,6 +326,11 @@ impl NotificationStore {
         pane_ids.iter().map(|id| self.pane_unread(*id)).sum()
     }
 
+    /// Total unread count across all panes.
+    pub fn total_unread(&self) -> usize {
+        self.pane_states.values().map(|s| s.unread_count).sum()
+    }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-notify/src/lib.rs:332]: Fixed. Added `total_unread_across_panes` test that pushes notifications across two panes, verifies the sum (3), marks one pane read (→ 1), then removes the other pane (→ 0).

File: crates/amux-app/src/main.rs [hunk 13]
```diff
@@ -2677,17 +2772,78 @@ impl AmuxApp {
...
 
-            if pane_id == focused {
-                // Focused pane: still record but mark as read (flash only, no ring)
+            let source_str = source.as_str();
+
+            // Three-tier notification delivery (matching cmux):
+            // 1. App unfocused → system toast + custom command + unread
+            // 2. App focused, different pane → in-app sound + custom command + unread
+            // 3. App focused, same pane → mark read (flash only, no ring)
+            if !self.app_focused {
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-app/src/main.rs]: Fixed. Extracted `deliver_notification()` helper method that implements the three-tier logic once. Both `drain_notifications()` and the IPC `notify.send` handler now call it.

File: crates/amux-app/src/main.rs [hunk 14]
```diff
@@ -1253,6 +1323,31 @@ impl eframe::App for AmuxApp {
...
+                .position(|ws| ws.id == action.workspace_id)
+            {
+                self.active_workspace_idx = idx;
+                let ws = &mut self.workspaces[idx];
+                if ws.tree.iter_panes().contains(&action.pane_id) {
+                    ws.focused_pane = action.pane_id;
+                }
+                self.notifications.mark_pane_read(action.pane_id);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (daveowenatl) [reply crates/amux-app/src/main.rs]: Fixed. Moved `drain_actions()` processing before `drain_notifications()` so notification clicks navigate to the correct workspace/pane before new notifications are processed.

**Comment by daveowenatl** [comment]
## CodeRabbit review body findings (latest review)

**1. `main.rs:2775-2815` — Don't suppress the last-focused pane when the window is unfocused (Critical)**
Already fixed in commit 6ceed02. The code now checks `!self.app_focused` first (line 2781) before checking `pane_id == focused`, so unfocused-window notifications always take the Tier 1 (system toast) path.

**2. `main.rs:4572-4595` — `notify.send` still bypasses the three-tier delivery rules (Major)**
Already fixed in commit 6ceed02. The IPC handler now uses the full three-tier `app_focused + pane_id == focused` classification. In today's commit (7baa61a) both paths were further consolidated into a single `deliver_notification()` helper to prevent future divergence.

**Comment by daveowenatl** [comment]
## CI fix

All three CI jobs (ubuntu, macOS, Windows) were failing — ubuntu failed on `alsa-sys` build (missing `libasound2-dev` headers for the `rodio` audio crate), which caused macOS and Windows to be cancelled.

**Fix:** Added a step to install `libasound2-dev` and `libudev-dev` on Linux before the Rust toolchain setup in `.github/workflows/ci.yml`.

All review comments from previous rounds remain addressed — no new unaddressed findings.
